File: web/src/engine/domain/equipment.ts
// Portage de Domain/Equipment.swift.
import type { BikeType, RaceFormat, RacePriority, Sport } from './types';

export type StrengthAccess = 'gym' | 'homeWeights' | 'bodyweightOnly' | 'none';

/** État du matériel disponible à un instant T. */
export interface Equipment {
  hasBike: boolean;
  bikeType?: BikeType;
  bikeWeightKg?: number;
  hasAeroBars: boolean;
  hasPowerMeter: boolean;
  hasSmartTrainer: boolean;
  poolAccess: boolean;
  openWaterAccess: boolean;
  hasWetsuit: boolean;
  hasDrylandCords: boolean;
  runOutdoor: boolean;
  hasTreadmill: boolean;
  hasTrack: boolean;
  strengthAccess: StrengthAccess;
}

export function makeEquipment(p: Partial<Equipment> = {}): Equipment {
  return {
    hasBike: false,
    hasAeroBars: false,
    hasPowerMeter: false,
    hasSmartTrainer: false,
    poolAccess: false,
    openWaterAccess: false,
    hasWetsuit: false,
    hasDrylandCords: false,
    runOutdoor: true,
    hasTreadmill: false,
    hasTrack: false,
    strengthAccess: 'bodyweightOnly',
    ...p,
  };
}

/** Le sport peut-il être pratiqué en l'état ? */
export function canPractice(e: Equipment, sport: Sport): boolean {
  switch (sport) {
    case 'swim': return e.poolAccess || e.openWaterAccess;
    case 'bike': return e.hasBike;
    case 'run': return e.runOutdoor || e.hasTreadmill || e.hasTrack;
    case 'strength': return e.strengthAccess !== 'none';
    case 'brick': return e.hasBike && (e.runOutdoor || e.hasTreadmill);
  }
}

export interface Race {
  id: string;
  date: Date;
  format: RaceFormat;
  priority: RacePriority;
  title: string;
  /** Objectif de temps total (secondes). Pilote l'ambition du plan et le pacing. */
  goalTimeSeconds?: number;
}

let raceCounter = 0;
export function makeRace(p: Omit<Race, 'id'> & { id?: string }): Race {
  return { id: p.id ?? `race-${++raceCounter}-${Date.now()}`, ...p };
}

File: Packages/TriathlonEngine/Sources/TriathlonEngine/Domain/Equipment.swift
import Foundation

/// État du matériel disponible à un instant T (historisé côté persistance).
public struct Equipment: Codable, Sendable, Equatable {
    public var hasBike: Bool
    public var bikeType: BikeType?
    public var bikeWeightKg: Double?
    public var hasAeroBars: Bool
    public var hasPowerMeter: Bool
    public var hasSmartTrainer: Bool

    public var poolAccess: Bool
    public var openWaterAccess: Bool
    public var hasWetsuit: Bool
    public var hasDrylandCords: Bool     // élastiques de traction à sec

    public var runOutdoor: Bool
    public var hasTreadmill: Bool
    public var hasTrack: Bool

    public enum StrengthAccess: String, Codable, Sendable {
        case gym, homeWeights, bodyweightOnly, none
    }
    public var strengthAccess: StrengthAccess

    public init(
        hasBike: Bool = false,
        bikeType: BikeType? = nil,
        bikeWeightKg: Double? = nil,
        hasAeroBars: Bool = false,
        hasPowerMeter: Bool = false,
        hasSmartTrainer: Bool = false,
        poolAccess: Bool = false,
        openWaterAccess: Bool = false,
        hasWetsuit: Bool = false,
        hasDrylandCords: Bool = false,
        runOutdoor: Bool = true,
        hasTreadmill: Bool = false,
        hasTrack: Bool = false,
        strengthAccess: StrengthAccess = .bodyweightOnly
    ) {
        self.hasBike = hasBike
        self.bikeType = bikeType
        self.bikeWeightKg = bikeWeightKg
        self.hasAeroBars = hasAeroBars
        self.hasPowerMeter = hasPowerMeter
        self.hasSmartTrainer = hasSmartTrainer
        self.poolAccess = poolAccess
        self.openWaterAccess = openWaterAccess
        self.hasWetsuit = hasWetsuit
        self.hasDrylandCords = hasDrylandCords
        self.runOutdoor = runOutdoor
        self.hasTreadmill = hasTreadmill
        self.hasTrack = hasTrack
        self.strengthAccess = strengthAccess
    }

    /// Le sport peut-il être pratiqué en l'état ?
    public func canPractice(_ sport: Sport) -> Bool {
        switch sport {
        case .swim: return poolAccess || openWaterAccess
        case .bike: return hasBike
        case .run: return runOutdoor || hasTreadmill || hasTrack
        case .strength: return strengthAccess != .none
        case .brick: return hasBike && (runOutdoor || hasTreadmill)
        }
    }
}

public struct Race: Codable, Sendable, Equatable, Identifiable {
    public var id: UUID
    public var date: Date
    public var format: RaceFormat
    public var priority: RacePriority
    public var title: String

    public init(id: UUID = UUID(), date: Date, format: RaceFormat, priority: RacePriority, title: String) {
        self.id = id
        self.date = date
        self.format = format
        self.priority = priority
        self.title = title
    }
}
